### PR TITLE
PWN-3096 - make error handling from FeeRelayer

### PR DESCRIPTION
**Pull request by kabramovp2p**
https://p2pvalidator.atlassian.net/browse/PWN-3096

[ Made error handling from FeeRelayer and pass this error to OrcaSwapPresenter to show error on UI ]

Screenshot 1 | Screenshot 2 | 
:------------:|:-------------:
Android Emulator - Pixel_5_API_30:5554 2022-03-24 16-53-42 | Android Emulator - Pixel_5_API_30:5554 2022-03-24 16-55-29

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorConverter.kt [hunk 1]
```diff
@@ -0,0 +1,97 @@
+package org.p2p.wallet.infrastructure.network.feerelayer
+
+import org.p2p.wallet.utils.emptyString
+import java.util.Locale
+
+object ErrorConverter {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorConverter.kt:6]: FeeRelayerErrorMapper?

File: app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorConverter.kt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    private val escapePrefixes = listOf(
+        programFailed,
+        programError,
+        "Transfer: "
+    )
+
+    fun fromFeeRelayer(code: Int, rawError: String): FeeRelayerError {
+        val logRegex = Regex("\"(?:Program|Transfer:) [^\"]+\"")
+        val matches = logRegex.findAll(rawError)
+        val logs = matches.map { match ->
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorConverter.kt:28]: maybe flatMap can be used here?

File: app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorConverter.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    private fun String?.escapePrefixes(prefixes: List<String>): String? {
+        var resultMessage = this ?: return null
+        prefixes.forEach {
+            resultMessage = resultMessage.replace(it, emptyString())
+        }
+        return resultMessage
+    }
+
+    private fun String.containsAnyOf(keywords: List<String>): Boolean {
+        for (keyword in keywords) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorConverter.kt:92]: maybe can be rewritten as
`keywords.any { contains(it) }`?

File: app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorTypeConverter.kt [hunk 4]
```diff
@@ -1,13 +1,22 @@
...
 object ErrorTypeConverter {
 
-    fun fromFeeRelayer(errorType: FeeRelayerErrorType): ErrorCode {
-        // TODO: Add implementation
-        return ErrorCode.SERVER_ERROR
+    fun fromFeeRelayer(error: FeeRelayerError): ErrorCode {
+        return when (error.type) {
+            FeeRelayerErrorType.SLIPPAGE_LIMIT -> ErrorCode.SLIPPAGE_LIMIT
+            FeeRelayerErrorType.INSUFFICIENT_FUNDS -> ErrorCode.INSUFFICIENT_FUNDS
+            else -> when (error.code) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorTypeConverter.kt:14]: when inside when is not a good idea. can be extracted it to a method

File: app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorConverter.kt [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            log?.contains("exceeds desired slippage") == true -> FeeRelayerErrorType.SLIPPAGE_LIMIT
+            log?.contains("exceeded maximum number of instructions allowed") == true -> {
+                FeeRelayerErrorType.MAXIMUM_NUMBER_OF_INSTRUCTIONS_ALLOWED_EXCEEDED
+            }
+            else -> FeeRelayerErrorType.UNKNOWN
+        }
+    }
+
+    private fun List<String>.findFirstValidLog(prefixes: List<String>): String? {
+        forEach { log ->
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorConverter.kt:65]: `firstOrNull { it.containsAnyOf(prefixes) }`

File: app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorConverter.kt (hunk 3 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorConverter.kt:92]: and you can inline it in the call site

File: app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorConverter.kt [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        val currentLog = logs.findFirstValidLog(errorPrefixes)
+
+        val codeRegex = Regex("$codePrefix-?\\d+")
+        val feeRelayerCode = codeRegex.findAll(rawError)
+            .map {
+                it.groupValues
+            }.toList()
+            .flatten()
+            .firstOrNull()
+            .toString()
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/feerelayer/ErrorConverter.kt:42]: will cause "null" string if `firstOrNull` results null?

**Review by 1sakov — APPROVED** [review]
Approved, good job! @kabramovp2p

**Review by gslevinkov — APPROVED** [review]

**Comment by kabramovp2p** [comment]
will move to QA after #381 as it is related to FeeRelayer and should be checked only after this PR will be merged

**Review by davran1001 — APPROVED** [review]

**Review by kuznetsov — APPROVED** [review]

**Pull request merged**